tests/validators/test_dataclasses.py

```diff
@@ -1190,3 +1191,196 @@ def test_custom_dataclass_names():
+        (
+            'always',
+            DuplicateDifferentSlots(a='hello', b=True),
+            Err('should be a dictionary or an instance of FooDataclass'),
+        ),
+        # revalidate_instances='subclass-instances'
+        ('subclass-instances', {'a': 'hello', 'b': True}, {'a': 'hello', 'b': True}),
+        ('subclass-instances', FooDataclassSlots(a='hello', b=True), {'a': 'hello', 'b': True}),
+        ('subclass-instances', FooDataclassSlots(a=b'hello', b='true'), {'a': b'hello', 'b': 'true'}),
+        ('subclass-instances', FooDataclassSameSlots(a='hello', b=True), {'a': 'hello', 'b': True}),
+        ('subclass-instances', FooDataclassSameSlots(a=b'hello', b='true'), {'a': 'hello', 'b': True}),
+        ('subclass-instances', FooDataclassMoreSlots(a='hello', b=True, c='more'), Err('Unexpected keyword argument')),
+        (
+            'subclass-instances',
+            DuplicateDifferentSlots(a='hello', b=True),
+            Err('dictionary or an instance of FooDataclass'),
+        ),
+        # revalidate_instances='never'
+        ('never', {'a': 'hello', 'b': True}, {'a': 'hello', 'b': True}),
+        ('never', FooDataclassSlots(a='hello', b=True), {'a': 'hello', 'b': True}),
+        ('never', FooDataclassSameSlots(a='hello', b=True), {'a': 'hello', 'b': True}),
+        ('never', FooDataclassMoreSlots(a='hello', b=True, c='more'), {'a': 'hello', 'b': True, 'c': 'more'}),
+        ('never', FooDataclassMoreSlots(a='hello', b='wrong', c='more'), {'a': 'hello', 'b': 'wrong', 'c': 'more'}),
+        (
+            'never',
+            DuplicateDifferentSlots(a='hello', b=True),
+            Err('should be a dictionary or an instance of FooDataclass'),
+        ),
+    ],
+)
+@pytest.mark.skipif(sys.version_info < (3, 10), reason='slots are only supported for dataclasses in Python >= 3.10')
+def test_slots_dataclass_subclass(revalidate_instances, input_value, expected):
+    schema = core_schema.dataclass_schema(
+        FooDataclassSlots,
+        core_schema.dataclass_args_schema(
+            'FooDataclass',
+            [
+                core_schema.dataclass_field(name='a', schema=core_schema.str_schema()),
+                core_schema.dataclass_field(name='b', schema=core_schema.bool_schema()),
+            ],
+            extra_behavior='forbid',
+        ),
+        revalidate_instances=revalidate_instances,
+        slots=['a', 'b'],
+    )
+    v = SchemaValidator(schema)
+
+    if isinstance(expected, Err):
+        with pytest.raises(ValidationError, match=expected.message) as exc_info:
+            print(v.validate_python(input_value))
+
+        # debug(exc_info.value.errors(include_url=False))
+        if expected.errors is not None:
+            assert exc_info.value.errors(include_url=False) == expected.errors
+    else:
+        dc = v.validate_python(input_value)
+        assert dataclasses.is_dataclass(dc)
+        assert dataclasses.asdict(dc) == expected
```

Comment:
Here's a test that's failing. I think we should keep the test for the validator part, and it's passing there, but it's failing for serialization and I think we need to fix that:
```python

def test_slots_mixed() -> None:
    @dataclasses.dataclass(slots=True)
    class Model:
        x: int

    @dataclasses.dataclass
    class SubModel(Model):
        y: int

    schema = core_schema.dataclass_schema(
        SubModel,
        core_schema.dataclass_args_schema(
            'SubModel', [
                core_schema.dataclass_field(name='x', schema=core_schema.int_schema()),
                core_schema.dataclass_field(name='y', schema=core_schema.int_schema()),
            ]
        ),
        slots=['x'],
    )
    v = SchemaValidator(schema)
    dc = v.validate_python({'x': 1, 'y': 2})
    assert dc.x == 1
    assert dc.y == 2

    from pydantic_core import SchemaSerializer
    s = SchemaSerializer(schema)
    assert s.to_python(dc) == {'x': 1, 'y': 2}   # fail; currently it gets {'y': 2}
```

Reply:
I think this is now fixed - ended up having to significantly refactor how dataclasses are serialized.